test/unit/util/datatype_black.h

```diff
@@ -46,172 +46,188 @@ class DatatypeBlack : public CxxTest::TestSuite {
                          << "  is " << colorsType.mkGroundTerm() << endl;
     TS_ASSERT(colorsType.mkGroundTerm().getType() == colorsType);
   }
 
   void testNat() {
-    Datatype nat(d_em, "nat");
+    DType nat("nat");
 
-    DatatypeConstructor succ("succ");
-    succ.addArg("pred", DatatypeSelfType());
+    std::shared_ptr<DTypeConstructor> succ =
+        std::make_shared<DTypeConstructor>("succ");
+    succ->addArgSelf("pred");
     nat.addConstructor(succ);
 
-    DatatypeConstructor zero("zero");
+    std::shared_ptr<DTypeConstructor> zero =
+        std::make_shared<DTypeConstructor>("zero");
     nat.addConstructor(zero);
 
     Debug("datatypes") << nat << std::endl;
-    DatatypeType natType = d_em->mkDatatypeType(nat);
+    TypeNode natType = d_nm->mkDatatypeType(nat);
     Debug("datatypes") << natType << std::endl;
 
-    Expr ctor = natType.getDatatype()[1].getConstructor();
-    Expr apply = d_em->mkExpr(kind::APPLY_CONSTRUCTOR, ctor);
+    Node ctor = natType.getDType()[1].getConstructor();
+    Node apply = d_nm->mkNode(kind::APPLY_CONSTRUCTOR, ctor);
     Debug("datatypes") << apply << std::endl;
 
-    TS_ASSERT(! natType.getDatatype().isParametric());
-    TS_ASSERT(! natType.getDatatype().isFinite());
-    TS_ASSERT(natType.getDatatype().getCardinality().compare(Cardinality::INTEGERS) == Cardinality::EQUAL);
-    TS_ASSERT(natType.getDatatype().isWellFounded());
-    Debug("groundterms") << "ground term of " << natType.getDatatype().getName() << endl
+    TS_ASSERT(!natType.getDType().isParametric());
+    TS_ASSERT(!natType.getDType().isFinite());
+    TS_ASSERT(natType.getDType().getCardinality().compare(Cardinality::INTEGERS)
+              == Cardinality::EQUAL);
+    TS_ASSERT(natType.getDType().isWellFounded());
+    Debug("groundterms") << "ground term of " << natType.getDType().getName()
+                         << endl
                          << "  is " << natType.mkGroundTerm() << endl;
     TS_ASSERT(natType.mkGroundTerm().getType() == natType);
   }
 
   void testTree() {
-    Datatype tree(d_em, "tree");
-    Type integerType = d_em->integerType();
+    DType tree("tree");
+    TypeNode integerType = d_nm->integerType();
 
-    DatatypeConstructor node("node");
-    node.addArg("left", DatatypeSelfType());
-    node.addArg("right", DatatypeSelfType());
+    std::shared_ptr<DTypeConstructor> node =
+        std::make_shared<DTypeConstructor>("node");
+    node->addArgSelf("left");
+    node->addArgSelf("right");
     tree.addConstructor(node);
 
-    DatatypeConstructor leaf("leaf");
-    leaf.addArg("leaf", integerType);
+    std::shared_ptr<DTypeConstructor> leaf =
+        std::make_shared<DTypeConstructor>("leaf");
+    leaf->addArg("leaf", integerType);
     tree.addConstructor(leaf);
 
     Debug("datatypes") << tree << std::endl;
-    DatatypeType treeType = d_em->mkDatatypeType(tree);
+    TypeNode treeType = d_nm->mkDatatypeType(tree);
     Debug("datatypes") << treeType << std::endl;
 
```

Comment:
What about the tests for `getConstructor()`? Don't we have something similar at the Node level?

Reply:
We don't have any equivalent tests for constructor.  We have a DTypeConstructor, although its tests are grouped with DType in this unit test.